src/CrateTokenV1.sol

```diff
@@ -54,100 +44,123 @@ contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard {
         artistFeeDestination = _artistAddress;
         protocolFeeDestination = _protocolAddress;
         uniswapV2Router02 = _uniswapV2Router02;
-        bondingCurveActive = true;
-
+        phase = Phase.CROWDFUND;
         tokensInCurve = MAX_CURVE_SUPPLY;
         songURI = _songURI;
         _approve(address(this), uniswapV2Router02, MAX_SUPPLY);
     }
 
-    function buyWithEth(uint256 minTokensReceivable) external payable {
-        // Take fees out of ETH, then see how many tokens you can buy with the remaining amount.
-        require(bondingCurveActive, "Bonding curve ended");
-        require(tokensInCurve > 0, "Bonding curve is sold out");
+    /// PUBLIC ///
 
+    function buyWithEth(uint256 minTokensReceivable) external payable {
         uint256 preFee =
             (msg.value * (CRATE_FEE_PERCENT + ARTIST_FEE_PERCENT)) / (CRATE_FEE_PERCENT + ARTIST_FEE_PERCENT + 1 ether);
-        uint256 netValue = msg.value - preFee;
+        uint256 netValue = ((msg.value - preFee) * 999) / 1000;
         uint256 tokensToBuy = estimateMaxPurchase(netValue);
-
-        require(tokensToBuy > 0, "Not enough ETH provided to buy tokens.");
-        require(tokensToBuy >= minTokensReceivable, "Slippage tolerance exceeded.");
-
-        //If you sent enough eth, then buy tokens.
+        if (tokensToBuy == 0) revert Zero();
+        if (tokensToBuy < minTokensReceivable) revert SlippageToleranceExceeded();
         buy(tokensToBuy);
     }
 
     function buy(uint256 _amount) public payable nonReentrant {
-        address sender = LibMulticaller.sender();
-
-        require(bondingCurveActive, "Bonding curve ended");
-        require(_amount <= tokensInCurve, "Not enough tokens in the bonding curve.");
+        if (_amount > tokensInCurve) revert InsufficientTokens();
         if (tokensInCurve >= 10 ** decimals()) {
-            require(_amount >= 10 ** decimals(), "Cannot buy less than 1 token from the bonding curve.");
+            if (_amount < 10 ** decimals()) revert MustBuyAtLeastOneToken();
         } else {
             _amount = tokensInCurve;
         }
 
-        uint256 price = getBuyPrice(_amount);
-        uint256 crateFee = (price * CRATE_FEE_PERCENT) / 1 ether;
-        uint256 artistFee = (price * ARTIST_FEE_PERCENT) / 1 ether;
-        uint256 totalPayment = price + crateFee + artistFee;
-        artistFees += artistFee;
-
-        require(msg.value >= totalPayment, "Not enough Ether to complete purchase.");
-        // Calculate the minimum amount of tokens that should be received based on the slippage tolerance
-        uint256 minTokens = _amount - ((_amount * SLIPPAGE_TOLERANCE) / 10_000);
-
-        require(balanceOf(address(this)) >= minTokens, "Slippage tolerance exceeded.");
+        uint256 totalPayment;
+        uint256 crateFee;
+        uint256 artistFee;
+
+        if (phase == Phase.CROWDFUND) {
```

Comment:
I am worried about this big code block from line 77 to 101.

There is some repetitive logic, and I am typically not a fan of nested `if` blocks.

Can we refactor this block into cleaner modules/functions/variables, and use the [Return Early pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8) instead of nested if-elses?

Reply:
We should also emit an event when transitioning from phases Crowdfund -> bonding curve